internal/grpc/services/usershareprovider/usershareprovider.go

```diff
@@ -497,40 +506,181 @@ func (s *service) GetReceivedShare(ctx context.Context, req *collaboration.GetRe
+			availableMountpoint, err := GetAvailableMountpoint(ctx, gatewayClient,
+				resourceStat.GetInfo().GetId(),
+				requestedMountpoint,
+			)
+			if err != nil {
+				return &collaboration.UpdateReceivedShareResponse{
+					Status: status.NewInternal(ctx, err.Error()),
+				}, nil
+			}
+
+			if !slices.Contains(req.GetUpdateMask().GetPaths(), _fieldMaskPathMountPoint) {
+				req.GetUpdateMask().Paths = append(req.GetUpdateMask().GetPaths(), _fieldMaskPathMountPoint)
+			}
+
+			req.GetShare().MountPoint = &provider.Reference{
+				Path: availableMountpoint,
+			}
+		}
+	}
+
 	var uid userpb.UserId
 	_ = utils.ReadJSONFromOpaque(req.Opaque, "userid", &uid)
-	share, err := s.sm.UpdateReceivedShare(ctx, req.Share, req.UpdateMask, &uid)
+	updatedShare, err := s.sm.UpdateReceivedShare(ctx, req.Share, req.UpdateMask, &uid)
 	if err != nil {
 		return &collaboration.UpdateReceivedShareResponse{
 			Status: status.NewInternal(ctx, "error updating received share"),
 		}, nil
 	}
 
-	res := &collaboration.UpdateReceivedShareResponse{
+	return &collaboration.UpdateReceivedShareResponse{
 		Status: status.NewOK(ctx),
-		Share:  share,
+		Share:  updatedShare,
+	}, nil
+}
+
+// GetAvailableMountpoint returns a new or existing mountpoint
+func GetAvailableMountpoint(ctx context.Context, gwc gateway.GatewayAPIClient, id *provider.ResourceId, name string) (string, error) {
+	listReceivedSharesRes, err := gwc.ListReceivedShares(ctx, &collaboration.ListReceivedSharesRequest{})
+	if err != nil {
+		return "", errtypes.InternalError("grpc list received shares request failed")
 	}
-	return res, nil
+
+	if err := errtypes.NewErrtypeFromStatus(listReceivedSharesRes.GetStatus()); err != nil {
+		return "", err
+	}
+
+	// we need to sort the received shares by mount point in order to make things easier to evaluate.
+	base := filepath.Clean(name)
+	mount := base
+	existingMountpoint := ""
+	mountedShares := make([]string, 0, len(listReceivedSharesRes.GetShares()))
+	var pathExists bool
+
+	for _, s := range listReceivedSharesRes.GetShares() {
+		resourceIDEqual := utils.ResourceIDEqual(s.GetShare().GetResourceId(), id)
+
+		if resourceIDEqual && s.State == collaboration.ShareState_SHARE_STATE_ACCEPTED {
+			// a share to the resource already exists and is mounted, remembers the mount point
+			_, err := utils.GetResourceByID(ctx, s.GetShare().GetResourceId(), gwc)
+			if err == nil {
+				existingMountpoint = s.GetMountPoint().GetPath()
+			}
+		}
+
+		if s.State == collaboration.ShareState_SHARE_STATE_ACCEPTED {
+			// collect all accepted mount points
+			mountedShares = append(mountedShares, s.GetMountPoint().GetPath())
+			if s.GetMountPoint().GetPath() == mount {
+				// does the shared resource still exist?
+				_, err := utils.GetResourceByID(ctx, s.GetShare().GetResourceId(), gwc)
+				if err == nil {
+					pathExists = true
+				}
+				// TODO we could delete shares here if the stat returns code NOT FOUND ... but listening for file deletes would be better
```

Comment:
This `TODO` sounds urgent. Is it not?

Reply:
same, its something from the old code which is already in use, i would like to keep it!
https://github.com/cs3org/reva/blob/40abb344337b95e2af8f3384fbd73c87f4ffdb9a/internal/http/services/owncloud/ocs/handlers/apps/sharing/shares/pending.go#L162

Reply:
I think that (deleting shares for no longer existing resources) is a topic for a different PR. I'd rather prefer not to open that can of worms here. :worm: :worm: